test/1_stdlib/Collection.swift

```diff
@@ -202,5 +202,41 @@ func testIsEmptyFirstLast() {
 }
 testIsEmptyFirstLast()
 
+/// A `CollectionType` that vends just the default implementations for
+/// `CollectionType` methods.
+struct CollectionOnly<T: CollectionType> : CollectionType {
+  var base: T
+
+  var startIndex: T.Index {
+    return base.startIndex
+  }
+
+  var endIndex: T.Index {
+    return base.endIndex
+  }
+
+  func generate() -> T.Generator {
+    return base.generate()
+  }
+
+  subscript(position: T.Index) -> T.Generator.Element {
+    return base[position]
+  }
+}
+
+func testFirstPerformance() {
+  // accessing `first` should not perform duplicate work on lazy collections
+  let col_ = (0..<10).lazy.filter({ print("filter: \($0)"); return (2..<8).contains($0) })
```

Comment:
@gribozavr It's kind of hard to figure out what's going on with all the gyb, but it looks like `DefaultedForwardCollection` cannot, in fact, wrap any other collection. It's hard-coded to wrapping `MinimalForwardCollection`. So either I go back to having my own minimal collection wrapper and just use it to wrap `LoggingCollection`, or I use `DefaultedForwardCollection` with a side-effect-ful `filter` predicate. I'm not sure which is better (or I could change `MinimalForwardCollection` and `DefaultedForwardCollection` to actually take a sequence to wrap, but I'm not sure if there's any particular reason they don't do that already).